### PR TITLE
Fix substitution to add correct types to leaf nodes

### PR DESCRIPTION
**Pull request by oflatt**

### CODE REVIEW

File: tree_in_context/src/type_analysis.egg [hunk 1]
```diff
@@ -40,39 +40,89 @@
...
-        (HasArgType e ty)
-      )
+(rule ((= lhs (Switch pred (Cons branch rest)))
+       (HasArgType branch ty))
+      ((HasArgType lhs ty))
+      :ruleset type-analysis)
+;; demand with one fewer branches
+(rule ((= lhs (Switch pred (Cons branch rest))))
+      ((Switch pred rest))
+      :ruleset type-analysis)
```
Comment (clyben) [inline tree_in_context/src/type_analysis.egg:110]: Is this rule creating a new switch? Will this create uncessary terms?

File: tree_in_context/src/type_analysis.egg [hunk 2]
```diff
@@ -40,39 +40,89 @@
       ((panic "arg type mismatch"))
       :ruleset type-analysis)
 
-; Pushing expected arg types down
```
Comment (clyben) [inline tree_in_context/src/type_analysis.egg:43]: Why these pushing down rules no longer needed?

**Review by clyben — APPROVED** [review]
Looks good to me! Have two questions on type analysis.

File: tree_in_context/src/type_analysis.egg (hunk 2 above)
Comment (oflatt) [reply tree_in_context/src/type_analysis.egg:43]: We are now going bottom-up instead of top-down

**Pull request merged**